transform/util.go

```diff
@@ -133,6 +129,82 @@ func getJSONRaw(data []byte, path string, pathRequired bool) ([]byte, error) {
 	return result, nil
 }
 
+// setPath updates the value with properly formatted timestamp(s) and properly
+// handles array indexing
+func setJSONRaw(data, out []byte, path string) ([]byte, error) {
+	var err error
+	splitPath := strings.Split(path, ".")
+	numOfInserts := 0
+
+	for element, k := range splitPath {
+		arrayRefs := jsonPathRe.FindAllStringSubmatch(k, -1)
+		if arrayRefs != nil && len(arrayRefs) > 0 {
+			objKey := arrayRefs[0][1]      // the key
+			arrayKeyStr := arrayRefs[0][2] // the array index
+			if arrayKeyStr == "*" {
+				// ArrayEach setup
+				splitPath[element+numOfInserts] = objKey
+				beforePath := splitPath[:element+numOfInserts+1]
+				afterPath := strings.Join(splitPath[element+numOfInserts+1:], ".")
+				// use jsonparser.ArrayEach to count the number of items in the
+				// array
+				var arraySize int
+				_, err = jsonparser.ArrayEach(data, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
+					arraySize++
+				}, beforePath...)
+				if err != nil {
+					return nil, err
+				}
+
+				// setJSONRaw() the rest of path for each element in results
+				for i := 0; i < arraySize; i++ {
+					var newPath string
+					// iterate through each item in the array by replacing the
+					// wildcard with an int and joining the path back together
+					newArrayKey := strings.Join([]string{"[", strconv.Itoa(i), "]"}, "")
+					beforePathStr := strings.Join(beforePath, ".")
+					beforePathArrayKeyStr := strings.Join([]string{beforePathStr, newArrayKey}, "")
+					// if there's nothing that comes after the array index,
+					// don't join so that we avoid trailing cruft
+					if len(afterPath) > 0 {
+						newPath = strings.Join([]string{beforePathArrayKeyStr, afterPath}, ".")
+					} else {
+						newPath = beforePathArrayKeyStr
+					}
+					// now call the function, but this time with an array index
+					// instead of a wildcard
+					data, err = setJSONRaw(data, out, newPath)
+					if err != nil {
+						return nil, err
+					}
+				}
+				return data, nil
+			}
+			// if not a wildcard then piece that path back together with the
+			// array index as an entry in the splitPath slice
+			splitPath = makePathWithIndex(arrayKeyStr, objKey, splitPath, element, numOfInserts)
```

Comment:
In `get` we confirm that the arrayKeyStr is an integer before we pass to Set (checking Atoi() for error). I think if something like `foo[g].bar` is passed it could lead to weird behavior where `[g]` is interpreted as a key instead of a malformed array index and you get a "Key path not found" or something like that which might cause it to get set as a new element. Unless there's a reason not to I'd suggest making the check? Might even decorate with a more helpful error message if feeling fancy.